src/btree/row_srch.c

```diff
@@ -102,7 +102,18 @@ __wt_search_insert(
     match = skiphigh = skiplow = 0;
     ins = last_ins = NULL;
     for (i = WT_SKIP_MAXDEPTH - 1, insp = &ins_head->head[i]; i >= 0;) {
-        if ((ins = *insp) == NULL) {
+        /*
+         * The algorithm requires that the skip list insert pointer is only read once within the
+         * loop. Whilst the compiler can change the code in a way that it reads the insert pointer
+         * value from memory again in the following code.
+         *
+         * In addition, a CPU with weak memory ordering, such as ARM, may speculatively read an old
+         * value. It is not OK and the reason is explained in the following comment.
```

Comment:
The situation is actually worse than this comment is saying. While it may sound weird, without a barrier, there is nothing ensuring that when we dereference the skiplist pointer that the data we get back is newer than when the pointer was written. This includes the key and anything we return to the caller. In assembly this is less of an issue because arm (like all processors other than alpha) honor something called an "address dependency" to ensure that if you read an address that was provably written after the data it  points to (using something like WT_PUBLISH, or a CAS with at lease release semantics) then all reads from the address will happen after the load of the address itself. You can see this described briefly at the bottom of [this page of the arm docs](https://developer.arm.com/documentation/102336/0100/Memory-ordering).

And that would be great if we were writing assembly. The intent was that the C memory model would provide access to this (nearly) universal hardware guarantee using the "consume" memory ordering. Unfortunately compiler venders have found it nearly impossible to ensure that address dependencies from high level source code are always carried through correctly to the generated assembly, and all of them now treat it identically to "acquire", which thankfully is really cheap. While sometimes you may get lucky (and we probably have so far), it is really best not to rely on this unless we want to start auditing assembly before every release.

TL;DR any time you are loading a pointer stored in a concurrently updated data structure, which this code is doing, you almost certainly need acquire semantics or `WT_ORDERED_READ`. This would apply even if this code was for a linked list with a single pointer per node, rather than a skiplist with multiple pointers.

Reply:
got it.